[GSoC] Switch joints framework to using intermediate frames

tjstienstra opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
This PR changes the basic definition of a joint. Instead of using a `parent_axis` in combination with a `child_axis` to determine the intermediate frame and joint axis, it now uses a separate intermediate frame for both the parent and child as well as a separately defined joint axis. This follows mostly a general definition of a mobilizer described in [this paper](https://link.springer.com/article/10.1007/s11071-010-9717-3) by Seth, A., Sherman, M., Eastman, P., & Delp, S.

#### Other comments
This will also make it more easy to implement a `Mobilizer` class, for defining more advanced joints as described in the paper.

This PR will also fix #23913 in the joints part.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.mechanics
   * BREAKING: deprecated parent_axis and child_axis arguments for the joint classes
   * Implemented a joint axis in combination with intermediate frames for defining joints
   * BREAKING: replaced parent_joint_pos and child_joint_pos arguments for parent_point and child_point
<!-- END RELEASE NOTES -->

tjstienstra committed:
Switch joints framework to using intermediate frames

TJStienstra commented:
@brocksam @moorepants @sidhu1012 FYI

TJStienstra commented:
It has become a bit of a big change. Feel free to ask questions. Some points of discussion are probably:
1. Should `parent_axis` and `child_axis` become deprecated?
2. Should `_locate_joint_pos` and `_locate_joint_frame` always make sure that their entity is new, e.g. should `_locate_joint_pos` return `body.masscenter` if the `joint_pos` is `0`?
3. Should the deprecated part of the tests be updated to the new definition? Or only a small part and add separate tests for this new definition?

tjstienstra committed:
Fix failing code-quality test

tjstienstra committed:
Fix failing test and simplify joint orient methods

tjstienstra committed:
Fix failing doctest

TJStienstra commented:
So here are some of my own considerations on these discussion points:
1. Should `parent_aixs` and `child_axis` become deprecated?

Yes:
- While the principle is well documented, it sometimes still feels a bit a magic, which makes you need to find out what orientation the parent and child will actually have when the coordinate is for example `0`.
- Having two methods to define the same thing makes it more difficult to understand.
- A deprecation like this is now still easy to make, since there are only few users of the joints framework.

No:
- It is a easy automatic method to define a joint, which at first is very intuitive.

My personal conclusion would be to deprecate it together with its accompanying protected methods.

2. Should `_locate_joint_pos` and `_locate_joint_frame` always make sure that their entity is new, e.g. should `_locate_joint_pos` return `body.masscenter` if the `joint_pos` is `0`?

I would opt for not always creating a new entity, since it saves unnecessary points and frames. And we do not allow changing for example the joint position, because on changing the position the set velocities should also change and we cannot track these position updates.

3. Should the deprecated part of the tests be updated to the new definition? Or only a small part and add separate tests for this new definition?

Not sure what is usual within SymPy.

4. If someone does not specify the `joint_axis` should it be `parent.frame.x` or `parent_interframe.x`?

I would expect `parent_interframe.x` since that is the joint attachment. And it is also most probable that if someone did not specify the `joint_axis` (s)he won't specify the `parent_interframe` as well, meaning that it will still be `parent.frame.x`.

tjstienstra committed:
Add tests for intermediate frame and joint axis functionality

tjstienstra committed:
Fix PrismaticJoint bug

tjstienstra committed:
ÃUpdate tests to intermediate frame functionality

TJStienstra commented:
Since the PR now also supports the arguments `parent_joint_pos` and `child_joint_pos` to be `Point` objects, I would also like to deprecate this old name and use `parent_point` and `child_point`.

This is of course a next point of discussion.

tjstienstra committed:
Update documentation to include intermediate frames

tjstienstra committed:
Fix failing doctest

tjstienstra committed:
Switch parent_axis test to joint_axis

brocksam commented:
> 1. Should `parent_axis` and `child_axis` become deprecated?

Yes. The concepts of `parent_interframe` and `child_interframe` are much clearer. Providing a suitable `DeprecationWarning`, as you have done, is all that is needed here.

> 2. Should `_locate_joint_pos` and `_locate_joint_frame` always make sure that their entity is new, e.g. should `_locate_joint_pos` return `body.masscenter` if the `joint_pos` is `0`?

As it's not possible to change these after the initialisation of the instance, I think it's better to reuse `body.masscenter` and `body.frame` to save on instantiating what is to all intents and purposes a duplicate object.

> 3. Should the deprecated part of the tests be updated to the new definition? Or only a small part and add separate tests for this new definition?

Provided that `test_pinjoint_arbitrary_axis`, `test_pinjoint_pi` and `test_slidingjoint_arbitrary_axis` test the specific behaviour of `parent_axis` and `child_axis` to the level where we can be sure `parent_axis` and `child_axis` are working correctly, I think it's fine to update the other tests in place.

> 4. If someone does not specify the `joint_axis` should it be `parent.frame.x` or `parent_interframe.x`?

Ideally, it should be a requirement that `joint_axis` is specified, otherwise this potentially does something magically which will only be correct by chance in a small subset of cases. I understand that for backwards compatibility that previously there was a default argument here so enforcing it to be required breaks backwards compatibility. I'd suggest using `parent_interframe.x` but adding a warning that in a future version `joint_axis` will become a required argument.

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -224,45 +269,101 @@ def _orient_frames(self):
+            ax_mat = axis.to_matrix(self.child_interframe)
+        try:
+            self.parent_interframe.dcm(frame)  # Check if connected
+            int_frame = self.parent_interframe
```
Move to separate `else` in try-except block as this line won't be the one throwing the intended `ValueError`.

brocksam commented on doc/src/modules/physics/mechanics/joints.rst:
```diff
@@ -17,52 +17,68 @@ Joints in Physics/Mechanics
-   ...     child_joint_pos=-(child.frame.x + child.frame.y) / 10,
-   ...     parent_axis=parent.frame.x, child_axis=child.frame.x)
+   >>> joint = PinJoint(
+   ...     'hinge', parent, child, theta, omega,
```
Can we make `theta` and `omega` keyword arguments here? Otherwise it relies on the person reading the documentation having good understanding on `PinJoint`s constructor signature.

brocksam reviewed (commented):
Use of a mixin class will help avoid calling the `_express_axis` method with an invalid value for `axis`.

tjstienstra committed:
Move joint axis support to mixin

tjstienstra committed:
Deprecate parent_joint_pos and child_joint_pos

TJStienstra commented on sympy/physics/mechanics/joint.py:
```diff
@@ -224,45 +269,101 @@ def _orient_frames(self):
+                             'associated body.')
+        return ax
+
+    def _express_axis(self, axis, frame):
```
Thank you for the suggestion. Is the `JointAxisMixin` as you would expect? Did leave the `joint_axis` property with the subclasses themselves, so it is more clear in the documentation.

TJStienstra commented:
Still have to add a separate `warns_sympy_deprecation()`, checking if the user gets the warning and some of the other points listed [here](https://docs.sympy.org/dev/contributing/deprecations.html#how-to-deprecate-code)

tjstienstra committed:
Update docstrings

tjstienstra committed:
Add interframe and joint_axis tests

tjstienstra committed:
Fix deprecation docstring formatting

TJStienstra commented:
Any idea why the symengine test is failing? It passes on my pc.

TJStienstra commented:
It is related to #21988

tjstienstra committed:
Fix failing symengine test

tjstienstra committed:
Improve to_vector helper

TJStienstra commented on sympy/physics/mechanics/tests/test_joint.py:
```diff
@@ -150,12 +179,142 @@ def test_pin_joint_chaos_pendulum():
                                        (h/4 + lB/2)*omega)*A.x
 
 
+@XFAIL
```
@sidhu1012 I thought based on #22304 that this issue would be fixed, but it still fails when using the symengine. Could you give some clarification?

sidhu1012 replied:
This https://github.com/sympy/sympy/issues/22304#issuecomment-1085461259 says that the issue is fixed, earlier it wasn't sympy's error, I think using latest symengine version should fix this.

TJStienstra replied:
After [this commit](https://github.com/sympy/sympy/pull/23920/commits/30d833b1d3e5a0703ef2d7da6681d23f1fbd75f5) I've introduced Xfail, but before that it was still failing on the latest symengine version (I tested 0.9.2 local).

TJStienstra replied:
@sidhu1012 It seems it is still not fixed see [#22304 comment](https://github.com/sympy/sympy/issues/22304#issuecomment-1239321187)

TJStienstra commented on sympy/physics/mechanics/joint.py:
```diff
@@ -518,23 +703,25 @@ def _generate_speeds(self, speed):
+        self.child_interframe.set_ang_vel(self.parent_interframe, self.speeds[
+            0] * self.joint_axis.normalize())
 
     def _set_linear_velocity(self):
```
Currently the velocities are only set in the body frames not also in the intermediate frames. This means that these intermediate frames can still be translating with respect to their body, which should not be the case.
Either some extra `set_vel`'s should be added or `self.parent_point.set_vel(self.parent.frame, 0)` should be changed to `self.parent_point.set_vel(self.parent_interframe, 0)` (same for the child).

Which option would be best?

TJStienstra replied:
Here is a simple experiment:
```py
from sympy import pi
from sympy.physics.mechanics import Body, PrismaticJoint, ReferenceFrame

P, C = Body('P'), Body('C')
Pint, Cint = ReferenceFrame('Pint'), ReferenceFrame('Cint')
Pint.orient_axis(P.frame, P.frame.x, pi)
Cint.orient_axis(C.frame, C.frame.y, -pi / 2)
joint = PrismaticJoint('P', P, C, parent_interframe=Pint, child_interframe=Cint,
                       parent_point=P.x, child_point=-C.y, joint_axis=P.z)
P.masscenter.vel(joint.parent_interframe)
ValueError: Velocity of point P_masscenter has not been defined in ReferenceFrame Pint
```

TJStienstra replied:
We could also just add a `body.masscenter.set_vel(int_frame, 0)` in the interframe creation method.

moorepants commented:
Two main points:

1. I don't really understand from a users perspective how I create an "inter frame" what that means. I conceptually understand starting with a body that has a ref. frame and mass center and then adding a point and rotation axis vector relative to the body, then doing that for a second body, and then making the points coincident and the vectors colinear. What is the conceptual process now?
2. I don't' see how to create a joint axis that has an arbitrary direction on both the child and parent. How is that supported?

tjstienstra committed:
Add vector support to the intermediate frame arguments

TJStienstra commented:
@moorepants 
Hope point 1 is now more clear and for point 2 I have implemented vector support, which should make it easier.

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+body, i.e. ``parent_interframe`` and ``child_interframe``. This means that you
+can now fully define the joint attachment, consisting of a point and frame, for
+both bodies. And if a joint like the ``PinJoint`` has a specific joint axis,
+e.g. the axis around which the rotation occurs, than this axis can be specified
```
```suggestion
e.g. the axis around which the rotation occurs, then this axis can be specified
```

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+joints now use an intermediate frame argument for both the parent and the child
+body, i.e. ``parent_interframe`` and ``child_interframe``. This means that you
+can now fully define the joint attachment, consisting of a point and frame, for
+both bodies. And if a joint like the ``PinJoint`` has a specific joint axis,
```
```suggestion
both bodies. Furthermore, if a joint like the ``PinJoint`` has a specific joint axis,
```

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+can now fully define the joint attachment, consisting of a point and frame, for
+both bodies. And if a joint like the ``PinJoint`` has a specific joint axis,
+e.g. the axis around which the rotation occurs, than this axis can be specified
+using the ``joint_axis`` argument. A big advantage of this setup is that you can
+be more accurately define the transformation from parent body to child body.
```
```suggestion
using the ``joint_axis`` argument. An advantage of this setup is that one can
more accurately define the transformation from a parent body to a child body.
```

TJStienstra replied:
would say "the parent body to the child body"

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+using the ``joint_axis`` argument. A big advantage of this setup is that you can
+be more accurately define the transformation from parent body to child body.
+
+For example, suppose you want a ``PinJoint`` rotates the child body around the
```
```suggestion
For example, suppose you want a ``PinJoint`` that rotates the child body around the
```

brocksam commented on doc/src/explanation/active-deprecations.md:
````diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+```
+
+However if you liked the fact that the deprecated arguments aligned the frames
+for you, than you can still make use of this feature by providing vectors to
````
```suggestion
for you, then you can still make use of this feature by providing vectors to
```

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+
+However if you liked the fact that the deprecated arguments aligned the frames
+for you, than you can still make use of this feature by providing vectors to
+``parent_interframe`` and ``child_interframe``, which are than oriented such
```
```suggestion
``parent_interframe`` and ``child_interframe``, which are then oriented such
```

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+for you, than you can still make use of this feature by providing vectors to
+``parent_interframe`` and ``child_interframe``, which are than oriented such
+that the joint axis expressed in the intermediate frame is aligned with the
+given vector.
```
```suggestion
given vector:
```

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+
+The argument names for specifying the attachment points of a joint in
+``sympy.physics.mechanics`` , i.e. ``parent_joint_pos`` and ``child_joint_pos``,
+has been changed to ``parent_point`` and ``child_point``. Since these arguments
```
```suggestion
have been changed to ``parent_point`` and ``child_point``. Since these arguments
```

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,125 @@ SymPy deprecation warnings.
+(deprecated-mechanics-joint-pos)=
+### Change in joint attachment point argument
+
+The argument names for specifying the attachment points of a joint in
+``sympy.physics.mechanics`` , i.e. ``parent_joint_pos`` and ``child_joint_pos``,
+has been changed to ``parent_point`` and ``child_point``. Since these arguments
+can now also be ``Point`` objects, so they can be exactly the same as the
+``parent_point`` and ``child_point`` attributes.
```
```suggestion
The argument names for specifying the attachment points of a joint in
``sympy.physics.mechanics`` , i.e. ``parent_joint_pos`` and ``child_joint_pos``,
have been changed to ``parent_point`` and ``child_point``. This is because these
arguments can now also be ``Point`` objects, so they can be exactly the same as
the ``parent_point`` and ``child_point`` attributes.
```

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -44,18 +45,48 @@ class Joint(ABC):
-        and child are connected. The default value is the zero vector.
+    parent_point : Point or Vector, optional
+        Attachment point where the joint is fixed to the parent body. If a
+        vector is provided, than the attachment point is computed by adding the
```
```suggestion
        vector is provided, then the attachment point is computed by adding the
```

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -44,18 +45,48 @@ class Joint(ABC):
+        center.
+    child_point : Point or Vector, optional
+        Attachment point where the joint is fixed to the child body. If a
+        vector is provided, than the attachment point is computed by adding the
```
```suggestion
        vector is provided, then the attachment point is computed by adding the
```

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -44,18 +45,48 @@ class Joint(ABC):
+            For more information on this deprecation, see
+            :ref:`deprecated-mechanics-joint-axis`.
+    parent_interframe : ReferenceFrame, optional
+        Interframe of the joint fixed to the parent body, i.e. frame in the of
+        parent body with respect to which the joint is formulated.  If a Vector
```
This first sentence isn't clear, possible typo?

TJStienstra replied:
It should have been: `Interframe of the joint fixed to the parent body, i.e. frame in the parent body with respect to which the joint is formulated.`
However I think it could be made clearer by using one of the following:
1. Joint attachment frame of the parent body, i.e. frame fixed to the parent body with respect to which the joint transformation is formulated.
2. Intermediate frame of the parent body with respect to which the joint transformation is formulated.

Which one do you think is more clear? Or do you have a better suggestion?

brocksam replied:
I think #2 is clearest.

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -44,18 +45,48 @@ class Joint(ABC):
+        is provided then an interframe is created which aligns its X axis with
+        the given vector. The default value is the parent's own frame.
+    child_interframe : ReferenceFrame, optional
+        Interframe of the joint fixed to the child body, i.e. frame in the of
+        child body with respect to which the joint is formulated. If a Vector
```
This first sentence isn't clear, possible typo?

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -71,60 +102,38 @@ class Joint(ABC):
     child_axis : Vector
         The axis fixed in the child frame that represents the joint.
+    parent_interframe : ReferenceFrame
+        Interframe of the joint fixed to the parent body, i.e. frame in the of
+        parent body with respect to which the joint is formulated.
```
This first sentence isn't clear, possible typo?

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -71,60 +102,38 @@ class Joint(ABC):
+        Interframe of the joint fixed to the parent body, i.e. frame in the of
+        parent body with respect to which the joint is formulated.
+    child_interframe : ReferenceFrame
+        Interframe of the joint fixed to the child body, i.e. frame in the of
+        child body with respect to which the joint is formulated.
```
This first sentence isn't clear, possible typo?

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -71,60 +102,38 @@ class Joint(ABC):
-       * - ``-C.x-C.y-C.z``
-         - ``P.x+P.y+P.z``
-         - ``(P.x+P.y+P.z) × P.x``
+    When providing a vector as intermediate frame, a new intermediate frame is
+    created which aligns its X axis with the provided vector. This is done by
+    with a single fixed rotation around a rotation axis. This rotation axis is
+    determined by taking the cross product of the ``body.x`` axis with the
+    provided vector. In case of the provided vector being in the ``-body.x``
+    direction, than the rotation is done around the ``body.y`` axis.
```
```suggestion
    When providing a vector as the intermediate frame, a new intermediate frame
    is created which aligns its X axis with the provided vector. This is done
    with a single fixed rotation around a rotation axis. This rotation axis is
    determined by taking the cross product of the ``body.x`` axis with the
    provided vector. In the case where the provided vector is in the ``-body.x``
    direction, the rotation is done around the ``body.y`` axis.
```

tjstienstra committed:
Fix failing test

tjstienstra committed:
Fix typos

brocksam commented on doc/src/explanation/active-deprecations.md:
```diff
@@ -76,7 +76,126 @@ SymPy deprecation warnings.
+(deprecated-mechanics-joint-axis)=
+### New Joint intermediate frames
+
+The definition of the joint axis in the ``sympy.physics.mechanics`` has changed.
+Instead of using the arguments ``parent_axis`` and ``child_axis`` to
+automatically determine the joint axis and an intermediate reference frame. The
+joints now use an intermediate frame argument for both the parent and the child
+body, i.e. ``parent_interframe`` and ``child_interframe``. This means that you
+can now fully define the joint attachment, consisting of a point and frame, for
+both bodies. Furthermore, if a joint like the ``PinJoint`` has a specific joint
+axis, e.g. the axis around which the rotation occurs, then this axis can be
+specified using the ``joint_axis`` argument. An advantage of this setup is that
+one can more accurately define the transformation from the parent body to the
+child body.
```
```suggestion
The definition of the joint axis in the ``sympy.physics.mechanics`` has changed.
Instead of using the arguments ``parent_axis`` and ``child_axis`` to
automatically determine the joint axis and an intermediate reference frame, the
joints now use an intermediate frame argument for both the parent and the child
body, i.e. ``parent_interframe`` and ``child_interframe``. This means that you
can now fully define the joint attachment, consisting of a point and frame, for
both bodies. Furthermore, if a joint like the ``PinJoint`` has a specific joint
axis, e.g. the axis about which the rotation occurs, then this axis can be
specified using the ``joint_axis`` argument. An advantage of this setup is that
one can more accurately define the transformation from the parent body to the
child body.
```

TJStienstra replied:
Changed all "around"s to "about"s for consistency ([both are allowed](https://english.stackexchange.com/questions/45566/rotate-about-vs-rotate-around))

brocksam commented on doc/src/modules/physics/mechanics/joints.rst:
```diff
@@ -17,52 +17,76 @@ Joints in Physics/Mechanics
+``parent_interframe`` and ``child_interframe``. This means that if for example
+the joint axis is defined in the ``parent_interframe`` as $2\hat{p}_x +
+4\hat{p}_y + 3\hat{p}_z$, then this will also be $2\hat{c}_x + 4\hat{c}_y +
+3\hat{c}_z$ in the ``child_interframe``. Practically this means that in the case
+of the :class:`~.PinJoint`, also shown below, that the ``joint_axis`` is the
+axis of rotation. With the generalized coordinate :math:`\theta` as the angle of
+rotation and the generalized speed :math:`\omega` as the angular velocity.
```
```suggestion
3\hat{c}_z$ in the ``child_interframe``. Practically this means that in the case
of the :class:`~.PinJoint`, also shown below, the ``joint_axis`` is the axis of
rotation, with the generalized coordinate :math:`\theta` as the angle of rotation
and the generalized speed :math:`\omega` as the angular velocity.
```

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -737,15 +1068,20 @@ def _generate_speeds(self, speed):
+        self.child_interframe.set_ang_vel(self.parent_interframe, 0)
 
     def _set_linear_velocity(self):
+        self.child_point.set_pos(self.parent_point, self.coordinates[
+            0] * self.joint_axis.normalize())
```
I think the below is much more readable formatting.
```suggestion
        self.child_point.set_pos(
            self.parent_point,
            self.coordinates[0] * self.joint_axis.normalize()
        )
```

TJStienstra replied:
I'll make a separate variable `axis = self.joint_axis.normalize()`. Makes it more readable and keeps all set commands on a single line.

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -737,15 +1068,20 @@ def _generate_speeds(self, speed):
-        self.child_point.set_pos(self.parent_point, self.coordinates[0] * self.parent_axis.normalize())
-        self.child_point.set_vel(self.parent.frame, self.speeds[0] * self.parent_axis.normalize())
-        self.child.masscenter.set_vel(self.parent.frame, self.speeds[0] * self.parent_axis.normalize())
+        self.child_point.set_vel(self.parent.frame, self.speeds[
+            0] * self.joint_axis.normalize())
```
I think the below is much more readable formatting.
```suggestion
        self.child_point.set_vel(
            self.parent.frame,
            self.speeds[0] * self.joint_axis.normalize()
        )
```

brocksam commented on sympy/physics/mechanics/joint.py:
```diff
@@ -737,15 +1068,20 @@ def _generate_speeds(self, speed):
-        self.child.masscenter.set_vel(self.parent.frame, self.speeds[0] * self.parent_axis.normalize())
+        self.child_point.set_vel(self.parent.frame, self.speeds[
+            0] * self.joint_axis.normalize())
+        self.child.masscenter.set_vel(self.parent.frame, self.speeds[
+            0] * self.joint_axis.normalize())
```
I think the below is much more readable formatting.
```suggestion
        self.child.masscenter.set_vel(
            self.parent.frame,
            self.speeds[0] * self.joint_axis.normalize()
        )
```

tjstienstra committed:
Make test function naming more consistent

brocksam reviewed (commented):
Mostly suggested rephrasing of docstrings and minor code reformatting. Otherwise looks good.

TJStienstra commented on sympy/physics/mechanics/joint.py:
```diff
@@ -224,75 +276,226 @@ def _orient_frames(self):
+                frame_name = f'{frame.name}_int_frame'
+        angle = frame_axis.angle_between(align_axis)
+        rotation_axis = cross(frame_axis, align_axis)
+        if rotation_axis != Vector(0) or angle == pi:
+            if angle == pi:
+                rotation_axis = Joint._choose_rotation_axis(frame, align_axis)
+
+            int_frame = ReferenceFrame(frame_name)
+            int_frame.orient_axis(frame, rotation_axis, angle)
+            int_frame.set_ang_vel(frame, 0 * rotation_axis)
+            return int_frame
+        return frame
```
`rotation_axis ==0 and angle == 0` seems more correct to me.

tjstienstra committed:
Fixate interframes to bodies

tjstienstra committed:
Fix typos and refactor

TJStienstra commented:
@brocksam thank you for the many suggestions. I've addressed all of them. Some suggestions I've changed a little as mentioned, but most have been accepted as you've proposed.

brocksam commented on sympy/physics/mechanics/tests/test_joint.py:
```diff
@@ -603,7 +610,7 @@ def test_sliding_joint():
     assert C.ang_vel_in(P) == 0
     assert P.ang_vel_in(C) == 0
-
+test_sliding_joint()
```
Has this been left in by error?

TJStienstra replied:
Yes, that is my method of quickly testing a test. Will remove it.

tjstienstra committed:
Remove redundant test call

tjstienstra committed:
Add test combining joint axis with interframe axis arguments

tjstienstra committed:
Remove old unused protected methods

moorepants commented:
We discussed in our weekly meeting that this was ready. Merged.
